File: ml/models/test_model_api.js
import * as tf from '@tensorflow/tfjs';
import { predict } from '../../src/services/model_api.js';

(async () => {
  try {
    // Read input from stdin
    const input = await new Promise((resolve, reject) => {
      let data = '';
      process.stdin.on('data', chunk => data += chunk);
      process.stdin.on('end', () => resolve(JSON.parse(data)));
      process.stdin.on('error', reject);
    });

    // Convert input features to TensorFlow.js tensor
    const features = tf.tensor(input.features, [1, 1, 561], 'float32');

    // Call the predict function
    const result = await predict(features);

    // Output the result as JSON
    console.log(JSON.stringify(result));
  } catch (error) {
    console.error("Error:", error.message);
    process.exit(1);
  }
})();

File: ml/models/test_model_api.py
import subprocess
import json
import numpy as np

def test_predict():
    # Create dummy input with correct shape and dtype
    x = np.zeros((1, 1, 561), dtype=np.float32).tolist()  # Convert to list for JSON serialization

    # Call the JavaScript predict function via Node.js
    try:
        result = subprocess.run(
            ["node", "test_model_api.js"],  # JavaScript test script
            input=json.dumps({"features": x}),
            text=True,
            capture_output=True,
            check=True
        )
    except subprocess.CalledProcessError as e:
        print("Error calling Node.js script:", e.stderr)
        raise

    # Parse the JSON output from the Node.js script
    output = json.loads(result.stdout)

    # Basic sanity checks
    assert "label" in output, "Missing 'label' in prediction result"
    assert "confidence" in output, "Missing 'confidence' in prediction result"
    assert "probabilities" in output, "Missing 'probabilities' in prediction result"
    assert len(output["probabilities"]) == 6, "Probabilities should have 6 classes"

    print("Test passed successfully with output:", output)

if __name__ == "__main__":
    test_predict()
